---
# https://isc.sans.edu/forums/diary/Finding+Privilege+Escalation+Flaws+in+Linux/19207

- name: Ensure user _test is present
  user:
    name: _test
    state: present
  when: ansible_os_family != "Alpine"

- name: Alpine | Add _test user
  command: "adduser -D _test creates=/home/_test"
  become: yes
  ignore_errors: true
  when: ansible_os_family == "Alpine"
- name: Alpine | Unlock user
  command: "passwd -u _test"
  ignore_errors: true
  when: ansible_os_family == "Alpine"

- name: retrieve unix-privesc-check
  get_url:
    url: http://pentestmonkey.net/tools/unix-privesc-check/unix-privesc-check-1.4.tar.gz
    dest: /tmp/unix-privesc-check-1.4.tar.gz
    checksum: 'sha256:e1f85e7f69fabadd23756a3d76a0e3d864b21e28763b60ce2e6181290d1c1969'
    mode: '0644'
# offline from time to time...
  ignore_errors: true
  register: upc

- name: unarchive unix-privesc-check
  unarchive:
    src: /tmp/unix-privesc-check-1.4.tar.gz
    dest: /tmp/
    remote_src: true
  become: yes
  become_user: _test
  when: not (ansible_distribution == "Alpine" or ansible_distribution == "Debian") and upc is succeeded

- name: Alpine | unarchive unix-privesc-check
  command: tar xzf /tmp/unix-privesc-check-1.4.tar.gz
  args:
    chdir: /tmp
    creates: /tmp/unix-privesc-check-1.4
  become: yes
  become_user: _test
  when: (ansible_distribution == "Alpine" or ansible_distribution == "Debian") and upc is succeeded

- name: execute unix-privesc-check
  shell: "/tmp/unix-privesc-check-1.4/unix-privesc-check detailed | tee /tmp/unix-privesc-check.out"
  ignore_errors: true
  register: privcheck
  args:
    creates: /tmp/unix-privesc-check.out
  become: yes
  become_user: _test
  when: upc is succeeded

#- debug: var=privcheck.stdout_lines
- name: grep error+warning of unix-privesc
  shell: egrep --color=always '(ERROR|WARNING)' /tmp/unix-privesc-check.out | grep -v "Can't find full path of running program:" | sort | uniq
  changed_when: false
  ignore_errors: true
  when: upc is succeeded

## FIXME: cert issue on ubuntu trusty - hash
## FIXME: in some case, server returns 206 Partial Content
- name: retrieve linuxprivchecker
  get_url:
    url: https://raw.githubusercontent.com/sleventyeleven/linuxprivchecker/master/linuxprivchecker.py
    dest: /tmp/linuxprivchecker.py
    checksum: 'sha256:d5002d499e3b3653a9505b2781f51d6d38507816b7eb9e9367d2c59fadfcac8c'
    mode: '0644'
    validate_certs: false
  become: yes
  become_user: _test
  ignore_errors: true

- name: execute unix-privesc-check
  shell: "python /tmp/linuxprivchecker.py | tee /tmp/linuxprivchecker.out"
  register: linuxprivchecker
  args:
    creates: /tmp/linuxprivchecker.out
  become: yes
  become_user: _test
  ignore_errors: true

#- debug: var=linuxprivchecker.stdout_lines

- name: Ensure git is present
  package:
    name: git
    state: present

# http://www.kitploit.com/2017/11/kernelpop-kernel-privilege-escalation.html
# https://github.com/spencerdodd/kernelpop
- name: retrieve kernelpop
  git:
    repo: 'https://github.com/spencerdodd/kernelpop'
    dest: /home/_test/kernelpop
    version: HEAD
    update: no
  become: yes
  become_user: _test

- name: ensure python3 is present
  package:
    name: "{{ python3_pkg | default('python3') }}"
    state: present

- name: execute kernelpop
  shell: "python3 kernelpop.py -b | tee /tmp/kernelpop.out"
  args:
    chdir: /home/_test/kernelpop
    creates: /tmp/kernelpop.out
  become: yes
  become_user: _test
